### PR TITLE
CMCL-517: Support for input system in the samples

### PR DESCRIPTION
**Pull request by gaborkb**
### Purpose of this PR

https://jira.unity3d.com/browse/CMCL-517

When Input System is added to Cinemachine, the Cinemachine examples are throwing a lot of errors, because they are trying to read from the old backends (UnityEngine.Input). 

There is a guard that checks if the old backends are disabled (ENABLE_LEGACY_INPUT_MANAGER). I added this to all example scripts that use Input, and a Debug.Log that tells users how they can fix it.

### Testing status

- [ ] Added an automated test
- [ ] Passed all automated tests
- [x] Manually tested 

### Documentation status

- [x] Updated [CHANGELOG](https://keepachangelog.com/en/1.0.0/)
- [ ] Updated README (if applicable)
- [ ] Commented all public classes, properties, and methods
- [ ] Updated user documentation

### Samples status

- [ ] Re-packed Extras~/CinemachineExamples.unitypackage

### Comments to reviewers
I added a guard to AxisState too, but I think it is better if that error is actually displayed as it was before. See my review comment for more details.

### Package version

[Justification for updating either the patch, minor, or major version according to the [semantic versioning](https://semver.org/spec/v2.0.0.html) rules]

- [ ] Updated package version

### CODE REVIEW

File: Runtime/Core/AxisState.cs [hunk 1]
```diff
@@ -219,11 +219,13 @@ public bool Update(float deltaTime)
             
             if (m_InputAxisProvider != null)
                 m_InputAxisValue = m_InputAxisProvider.GetAxisValue(m_InputAxisIndex);
+#if ENABLE_LEGACY_INPUT_MANAGER
             else if (!string.IsNullOrEmpty(m_InputAxisName))
             {
                 try { m_InputAxisValue = CinemachineCore.GetInputAxis(m_InputAxisName); }
                 catch (ArgumentException e) { Debug.LogError(e.ToString()); }
             }
+#endif
```
Comment (gaborkb) [inline Runtime/Core/AxisState.cs:228]: This might not be desired, because it hides the fact that there is no correct input provider assigned!

Comment (gaborkb) [reply Runtime/Core/AxisState.cs:228]: @glabute

File: Samples~/Cinemachine Example Scenes/Shared/Scripts/InputSystemHelper.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        public static void EnableBackendsWarningMessage()
+        {
+            if (Time.realtimeSinceStartup - s_LastMessageTime > 5)
+            {
+                Debug.Log(
+                    "Old input backends are disabled. Cinemachine examples use UnityEngine.Input. " +
+                    "To enable input in our examples: Edit > Project Settings > Player, " +
+                    "set Active Input Handling to Both!\n" +
+                    "Note: This is probably because the Input System Package has been added to the project. To use " +
+                    "Cinemachine Virtual Cameras with Input System Package, add CinemachineInputProvider to them.");
```
Comment (gaborkb) [inline Samples~/Cinemachine Example Scenes/Shared/Scripts/InputSystemHelper.cs:18]: @JWalker0007

**Review by AntoineCharton — APPROVED** [review]
Awesome to add support for it and guidelines! Only thing check for the warning message the standardized name of the Built in or Classic input system.
https://docs.unity3d.com/Manual/com.unity.inputsystem.html

File: Samples~/Cinemachine Example Scenes/Shared/Scripts/InputSystemHelper.cs (hunk 2 above)
Comment (JWalker0007) [inline Samples~/Cinemachine Example Scenes/Shared/Scripts/InputSystemHelper.cs:18]: with the Input System Package......

**Review by JWalker0007 — APPROVED** [review]

**Review by glabute — APPROVED** [review]

**Review by glabute — CHANGES_REQUESTED** [review]
Wait, I take back my approval.  See AxisState.cs

File: Runtime/Core/AxisState.cs (hunk 1 above)
Comment (glabute) [reply Runtime/Core/AxisState.cs:228]: No, wait.  This change should be reverted, and instead change the default implementation of CinemachineCore.GetInputAxis

File: Runtime/Core/CinemachineCore.cs [hunk 3]
```diff
@@ -72,7 +72,11 @@ public static CinemachineCore Instance
         /// <summary>Delegate for overriding Unity's default input system.
         /// If you set this, then your delegate will be called instead of
         /// System.Input.GetAxis(axisName) whenever in-game user input is needed.</summary>
+#if ENABLE_LEGACY_INPUT_MANAGER
         public static AxisInputDelegate GetInputAxis = UnityEngine.Input.GetAxis;
+#else
+        public static AxisInputDelegate GetInputAxis = null;
```
Comment (glabute) [inline Runtime/Core/CinemachineCore.cs:78]: What happens if it gets called?  No one is checking for null.  I think it should be an empty function instead.

**Review by glabute — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,7 +8,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Bugfix: Negative Near Clip Plane value is kept when camera is orthographic.
 - Regression fix: could not change the projection of the main camera if a CM virtual camera is active.
 - Regression fix: Axis input was ignoring CM's IgnoreTimeScale setting.
-- 
+- Bugfix: Cinemachine assigns a default input controller delegate that returns 0 when Input System is present.
```
Comment (glabute) [inline CHANGELOG.md:11]: I think you mean " when the legacy input system is disabled"

File: Runtime/Core/CinemachineCore.cs [hunk 5]
```diff
@@ -72,7 +72,11 @@ public static CinemachineCore Instance
         /// <summary>Delegate for overriding Unity's default input system.
         /// If you set this, then your delegate will be called instead of
         /// System.Input.GetAxis(axisName) whenever in-game user input is needed.</summary>
+#if ENABLE_LEGACY_INPUT_MANAGER
         public static AxisInputDelegate GetInputAxis = UnityEngine.Input.GetAxis;
+#else
+        public static AxisInputDelegate GetInputAxis = delegate { return 0; };
```
Comment (glabute) [inline Runtime/Core/CinemachineCore.cs:78]: Did you test this?  I am unfamiliar with this syntax.

**Review by glabute — APPROVED** [review]

File: Runtime/Core/CinemachineCore.cs (hunk 5 above)
Comment (gaborkb) [reply Runtime/Core/CinemachineCore.cs:78]: Yes, I tested it.
What did you have in mind instead?

Comment (gaborkb) [reply Runtime/Core/CinemachineCore.cs:78]: I could also do this:
`public static AxisInputDelegate GetInputAxis = axisName => 0;`

But to me the delegate version was clearer.

Comment (gaborkb) [reply Runtime/Core/CinemachineCore.cs:78]: The OrbTransposer in CameraPositionTest tests this indirectly, when the legacy input system is disabled.

Comment (glabute) [reply Runtime/Core/CinemachineCore.cs:78]: No worries.  Like I said, I was unfamiliar with the syntax so wanted to make sure it was tested.

**Pull request merged**